piptools/scripts/compile.py

```diff
@@ -302,6 +304,17 @@ def cli(
         )
         emit_index_url = index
 
+    if allow_unsafe is None:
```

Comment:
This warning might affect most of the users and force them to use `--no-allow-unsafe`, which could annoy them. Would it make more sense to change the default value in the next major release instead (as we did it with `--build-isolation` in [5.0.0](https://github.com/jazzband/pip-tools/releases/tag/5.0.0))?

Reply:
The goal is to nudge users towards `--allow-unsafe`, not `--no-allow-unsafe`. That way, they become early adopters of the behavior planned for 6.0.0 in #1249.

If you prefer to wait for a major release, can we jump straight to #1249 and drop this PR? We'd be relying on the release notes rather than using a warning.

Reply:
I'd prefer to add `--no-allow-unsafe` option and set it on by default. I'm not sure about removing `--allow-unsafe/--no-allow-unsafe` options completely.

Reply:
The comment in #989 suggests differently:

> I'm starting to think it might make sense to deprecate --allow-unsafe and make it the default behavior, but that would be as a separate issue/PR for discussion.

To me, the suggestion to deprecate it also implies eventual removal. Deprecating to me sends the message "don't use this option".

---

My impression from the linked ticket is that the idea of "unsafe" is an outdated notion. Do you think there are still scenarios where these packages are unsafe? Is there still a use case?

---

> and set it on by default

In this PR or in a PR destined for the major release?

---

I can comply with the requests above, but have a good understanding of your position would be helpful.

Reply:
Actually, we haven't discussed yet the issue properly. My point is that we could:
1. introduce new option `--no-allow-unsafe`
2. update `--allow-unsafe` to true by default

There are some rare cases where a user might need `--no-allow-hash`, for example, when some dependency depends on an old pip version (for some reason) which could break the latest `pip-tools`.

Later we could deprecate `--no-allow-unsafe` and gather feedback from users.

Reply:
> In this PR or in a PR destined for the major release?

Any PR you like, but it should be definitely introduced in the major releases.

Reply:
> introduce new option --no-allow-unsafe

Done in #1265. This is backward compatible so can come before the next major release.

> There are some rare cases where a user might need --no-allow-hash, for example, when some dependency depends on an old pip version (for some reason) which could break the latest pip-tools.

How would this work? If the user installs a version of pip that is incompatible with pip-tools, won't pip-tools break? I'm a bit confused how this is a viable environment.

Reply:
> Done in #1265. This is backward compatible so can come before the next major release.

Thanks!

> How would this work? If the user installs a version of pip that is incompatible with pip-tools, won't pip-tools break? I'm a bit confused how this is a viable environment.

If a user installs a `requirements.txt` built with ` --no-allow-unsafe` then pip-tools wouldn't be broken.